crates/polars-utils/src/cache.rs

```diff
@@ -29,3 +38,197 @@ where
         }
     }
 }
+
+/// A fixed-size cache optimized for access speed. Does not implement LRU or use
+/// a full hash table due to cost, instead we assign two pseudorandom slots
+/// based on the hash of the key, and if both are full we evict the one that had
+/// the older last access.
+const MIN_FAST_FIXED_CACHE_SIZE: usize = 16;
+
+#[derive(Clone)]
+pub struct FastFixedCache<K, V> {
+    slots: Vec<CacheSlot<K, V>>,
+    access_ctr: Cell<u32>,
+    shift: u32,
+    hash_builder: RandomState,
+}
+
+impl<K: Hash + Eq, V> FastFixedCache<K, V> {
+    pub fn new(n: usize) -> Self {
+        let n = (n.max(MIN_FAST_FIXED_CACHE_SIZE)).next_power_of_two();
+        Self {
+            slots: zeroed_vec(n),
+            access_ctr: Cell::new(1),
+            shift: 64 - n.ilog2(),
```

Comment:
What is the rationale of this?

Reply:
Of what?

Reply:
The   `64 - n.iglo2`

Reply:
That computes the shift width we need so that `hash(x) >> self.shift` where `hash(x)` is an `u64` exactly has the range `[0, n)`.